diff --git a/scripts/release.py b/scripts/release.py
index c06bf529..3a7f27b8 100755
--- a/scripts/release.py
+++ b/scripts/release.py
@@ -59,6 +59,9 @@ proj_list = [ "lv_sim_eclipse_sdl"]
 ver_major = -1
 ver_minor = -1
 ver_patch = -1
+dev_ver_major = -1
+dev_ver_minor = -1
+dev_ver_patch = -1
 
 ver_str = ""
 dev_ver_str = ""
@@ -309,7 +312,7 @@ def publish_master():
     #Merge LVGL master to dev first to avoid "merge-to-dev.yml" running asynchronous
     os.chdir("./lvgl")
     cmd("git checkout dev")
-    cmd("git merge master -X theirs")
+    cmd("git merge master -X ours")     # Use ours to keep the version numbers
     cmd("git add .")
     cmd("git commit -am 'Merge master'", False)
     cmd("git push origin dev")
@@ -473,6 +476,11 @@ if __name__ == '__main__':
         exit(1)
      
     clone_repos()
+    get_lvgl_version("dev")
+    dev_ver_major = ver_major
+    dev_ver_minor = ver_minor
+    dev_ver_patch = ver_patch
+    
     get_lvgl_version("master")
     lvgl_prepare()
     lv_examples_prepare() 
@@ -483,16 +491,17 @@ if __name__ == '__main__':
     update_release_branches()
     publish_master()
  
+    projs_update()    
+    
     if dev_prepare == 'bugfix': 
         ver_patch = str(int(ver_patch) + 1)
-        ver_str = "v" + ver_major + "." + ver_minor + "." + ver_patch + "-dev"    
+        ver_str = "v" + str(ver_major) + "." + str(ver_minor) + "." + str(ver_patch) + "-dev"    
 
         print("Prepare bugfix version " + ver_str)
 
         lvgl_update_master_version()
         docs_update_latest_version()
 
-        get_lvgl_version("dev")
         dev_ver_str = "v" + ver_major + "." + ver_minor + "." + ver_patch + "-dev"
         merge_to_dev()
         
@@ -520,6 +529,5 @@ if __name__ == '__main__':
         docs_update_dev_version()
         publish_dev_and_master()
         
-    projs_update()    
     cleanup()
     
